### DISCUSSION TITLE
Django Asgi Daphne Channels

### DISCUSSION THREAD

**Discussion by hasserchatphon** [vercel/community#2902] [Help]
### Summary

Hello, 

Can someone tell me if it is possible to run websocket on versel ?

### Steps to Reproduce

`import os
from django.core.asgi import get_asgi_application
from django.urls import path
from channels.routing import ProtocolTypeRouter
from django.core.asgi import get_asgi_application
from channels.routing import ProtocolTypeRouter, URLRouter
from channels.security.websocket import AllowedHostsOriginValidator
import django

os.environ.setdefault("DJANGO_SETTINGS_MODULE", "vercel_app.settings")
django.setup()
from channels.auth import AuthMiddlewareStack
from rtc.consumer import SignalingConsumer
from rtc.async_consumer import VideoCallConsumer

"""
ASGI config for backupcall project.

It exposes the ASGI callable as a module-level variable named ``application``.

For more information on this file, see
https://docs.djangoproject.com/en/4.2/howto/deployment/asgi/
"""

# Initialize Django ASGI application early to ensure the AppRegistry
# is populated before importing code that may import ORM models.
# django_asgi_app = get_asgi_application()

app = ProtocolTypeRouter(
    {
        "http": get_asgi_application(),
        "https": get_asgi_application(),
        "websocket": AllowedHostsOriginValidator(
            AuthMiddlewareStack(
                URLRouter(
                    [
                        # path("rtc", SignalingConsumer.as_asgi()),
                        path("async-rtc", VideoCallConsumer.as_asgi())
                    ]
                )
            )
        ),
    }
)
{
  "builds": [
    {
      "src": "vercel_app/asgi.py",
      "use": "@vercel/python"
    }
  ],
  "routes": [
    {
      "src": "/(.*)",
      "dest": "vercel_app/asgi.py"
    }
  ]
}
`

**Comment by ylmzfurkan** [comment]
Hii @hasserchatphon 

Yes, it is possible to run WebSockets on Vercel. However, there are a few things to keep in mind.

First, Vercel Serverless Functions have a maximum execution duration of 10 seconds. This means that you cannot use a Serverless Function to maintain a persistent WebSocket connection.

Second, Vercel does not support WebSockets out of the box. You will need to use a third-party service, such as Pusher Channels or Redis, to provide WebSocket support.

Here is an example of how you would use Pusher Channels to run WebSockets on Vercel:

```
import os
from django.core.asgi import get_asgi_application
from django.urls import path
from channels.routing import ProtocolTypeRouter, URLRouter
from channels.security.websocket import AllowedHostsOriginValidator
import django

os.environ.setdefault("DJANGO_SETTINGS_MODULE", "vercel_app.settings")
django.setup()
from channels.auth import AuthMiddlewareStack
from rtc.consumer import SignalingConsumer
from rtc.async_consumer import VideoCallConsumer
from pusher import Pusher

pusher = Pusher(
    app_id="YOUR_APP_ID",
    key="YOUR_KEY",
    secret="YOUR_SECRET",
    cluster="YOUR_CLUSTER",
    ssl=True,
)

application = ProtocolTypeRouter({
    "http": get_asgi_application(),
    "websocket": AuthMiddlewareStack(
        URLRouter([
            path("ws/signaling/", SignalingConsumer.as_asgi()),
            path("ws/videocall/", VideoCallConsumer.as_asgi()),
        ])
    ),
})

@pusher.on("connection_established")
def on_connection_established(data):
    print("New connection established")

@pusher.on("channel_joined")
def on_channel_joined(data):
    print("Joined channel")

@pusher.on("event")
def on_event(data):
    print("Received event")

if __name__ == "__main__":
    pusher.connect()

```

This code will create a Vercel application that uses Pusher Channels to provide WebSocket support. The on_connection_established(), on_channel_joined(), and on_event() functions will be called when a new connection is established, a channel is joined, or an event is received, respectively.
